builder/openstack/openstack.go

```diff
@@ -30,188 +34,772 @@ const (
+		}
+		networkOctets[i] = byte(octet)
+	}
+	_, ipv4Net, err := net.ParseCIDR(entProvisionedNetwork.Cidr)
 	if err != nil {
-		return err
+		return fmt.Errorf("error while parsing cidr: %v", err)
+	}
+	if len(ipv4Net.Mask) != 4 {
+		return fmt.Errorf("mask is not correct length")
 	}
+	hostAddress := strings.Join(append(networkOctetStrings[:3], fmt.Sprint(entHost.LastOctet)), ".")
 
-	// generate vm name from ent
-	name := builder.generateVmName(competition, team, host, build)
+	// Create Security Group for host
+	builder.Logger.Log.Debugf("Creating security group %s", vmName)
+	opts := secgroups.CreateOpts{
+		Name:        vmName,
+		Description: fmt.Sprintf("Sec group for VM %s", vmName),
+	}
+	osSecGroup, err := secgroups.Create(computeClient, opts).Extract()
+	if err != nil {
+		return fmt.Errorf("failed to create security group: %v", err)
+	}
+	newVars := entHost.Vars
+	newVars["openstack_secgroup_id"] = osSecGroup.ID
+	err = entProvisionedHost.Update().SetVars(newVars).Exec(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to update provisioned host vars: %v", err)
+	}
+	ruleOpts := make([]secgroups.CreateRuleOpts, 0)
+	for _, port := range entHost.ExposedTCPPorts {
+		fromPort := 0
+		toPort := 0
+		portRange := strings.Split(port, "-")
+		if len(portRange) == 1 {
+			destPort, err := strconv.Atoi(portRange[0])
+			if err != nil {
+				return fmt.Errorf("could not convert TCP port \"%s\" to integer: %v", port, err)
+			}
+			fromPort = destPort
+			toPort = destPort
+		} else {
+			fromPort, err = strconv.Atoi(portRange[0])
+			if err != nil {
+				return fmt.Errorf("could not convert TCP port \"%s\" to integer: %v", port, err)
+			}
+			toPort, err = strconv.Atoi(portRange[1])
+			if err != nil {
+				return fmt.Errorf("could not convert TCP port \"%s\" to integer: %v", port, err)
+			}
+		}
+		opts := secgroups.CreateRuleOpts{
+			ParentGroupID: osSecGroup.ID,
+			FromPort:      fromPort,
+			ToPort:        toPort,
+			IPProtocol:    "TCP",
+			CIDR:          "0.0.0.0/0",
+		}
+		ruleOpts = append(ruleOpts, opts)
+	}
+	for _, port := range entHost.ExposedUDPPorts {
+		fromPort := 0
+		toPort := 0
+		portRange := strings.Split(port, "-")
+		if len(portRange) == 1 {
+			destPort, err := strconv.Atoi(portRange[0])
+			if err != nil {
+				return fmt.Errorf("could not convert TCP port \"%s\" to integer: %v", port, err)
+			}
+			fromPort = destPort
+			toPort = destPort
+		} else {
+			fromPort, err = strconv.Atoi(portRange[0])
+			if err != nil {
+				return fmt.Errorf("could not convert TCP port \"%s\" to integer: %v", port, err)
+			}
+			toPort, err = strconv.Atoi(portRange[1])
+			if err != nil {
+				return fmt.Errorf("could not convert TCP port \"%s\" to integer: %v", port, err)
+			}
+		}
+		opts := secgroups.CreateRuleOpts{
+			ParentGroupID: osSecGroup.ID,
+			FromPort:      fromPort,
+			ToPort:        toPort,
+			IPProtocol:    "UDP",
+			CIDR:          "0.0.0.0/0",
+		}
+		ruleOpts = append(ruleOpts, opts)
+	}
```

Comment:
Instead of just doing 0.0.0.0/0 by default I recommend using the same logic that is present in the AWS builder.
https://github.com/emperorcow/laforge/blob/main/builder/aws/aws.go#L369-L382

Reply:
Should be fixed in [3db9982](https://github.com/emperorcow/laforge/pull/30/commits/3db99821471b936585615480309cc36790f43900)